src/opf/sdpwrm.jl

```diff
@@ -0,0 +1,300 @@
+struct SDPOPF <: AbstractFormulation end
+
+"""
+    build_opf(SDPOPF, data, optimizer)
+
+Build an SDPOPF model.
+"""
+function build_opf(::Type{SDPOPF}, data::OPFData, optimizer;
+    T=Float64,
+)
+    # Grab some data
+    N, E, G = data.N, data.E, data.G
+    vmin, vmax = data.vmin, data.vmax
+    i0 = data.ref_bus
+    gs, bs = data.gs, data.bs
+    pd, qd = data.pd, data.qd
+    bus_arcs_fr, bus_arcs_to = data.bus_arcs_fr, data.bus_arcs_to
+    bus_gens = data.bus_gens
+    pgmin, pgmax = data.pgmin, data.pgmax
+    qgmin, qgmax = data.qgmin, data.qgmax
+    c0, c1, c2 = data.c0, data.c1, data.c2
+    gen_status = data.gen_status
+    bus_fr, bus_to = data.bus_fr, data.bus_to
+    gff, gft, gtf, gtt = data.gff, data.gft, data.gtf, data.gtt
+    bff, bft, btf, btt = data.bff, data.bft, data.btf, data.btt
+    dvamin, dvamax, smax = data.dvamin, data.dvamax, data.smax
+    branch_status = data.branch_status
+
+    wr_min, wr_max, wi_min, wi_max = compute_voltage_phasor_bounds(data)
+
+    model = JuMP.GenericModel{T}(optimizer)
+    model.ext[:opf_model] = SDPOPF
+
+    #
+    #   I. Variables
+    #
+    #=
+        Some generators and branches may be inactive, as indicated by `branch_status` and `gen_status`.
+        Primal variables associated to inactive components are still declared, and handled as follows:
+        * lower/upper bounds are set to zero
+        * constraint coefficients are set to zero
+    =#
+
+    # voltage magnitude and product
+    WR_start = zeros(N, N) + I
+    @variable(model, WR[i=1:N, j=1:N], Symmetric, start=WR_start[i,j])
+    @variable(model, WI[1:N, 1:N] in SkewSymmetricMatrixSpace(), start=0.0)
+
+    # Active and reactive dispatch
+    @variable(model, pg[g in 1:G])
+    @variable(model, qg[g in 1:G])
+    
+    # Directional branch flows
+    @variable(model, pf[e in 1:E])
+    @variable(model, qf[e in 1:E])
+    @variable(model, pt[e in 1:E])
+    @variable(model, qt[e in 1:E])
+
+    # 
+    #   II. Constraints
+    #
+
+    # Voltage magnitude bounds
+    set_lower_bound.([WR[i, i] for i in 1:N], vmin.^2)
+    set_upper_bound.([WR[i, i] for i in 1:N], vmax.^2)
+
+    # Voltage product bounds
+    set_lower_bound.([WR[bus_fr[e], bus_to[e]] for e in 1:E], wr_min)
+    set_upper_bound.([WR[bus_fr[e], bus_to[e]] for e in 1:E], wr_max)
+    set_lower_bound.([WI[sort([bus_fr[e], bus_to[e]])...] for e in 1:E], wi_min)
+    set_upper_bound.([WI[sort([bus_fr[e], bus_to[e]])...] for e in 1:E], wi_max)
```

Comment:
I believe those bounds are not needed in the SDP relaxation.

Reply:
Just like in the SOC formulation, they are not needed. They are implied by the change of variables, and are not implied by the other constraints.

`PowerModels` also applies these bounds: https://github.com/lanl-ansi/PowerModels.jl/blob/8824212b3d2f35f94f7dd72f37ff76e3419c478f/src/form/wrm.jl#L89-L95

Reply:
If you prefer to keep these, be sure to state them in the formulation that's in the docs